apps/sparo-lib/src/cli/commands/pull.ts

```diff
@@ -0,0 +1,95 @@
+import { inject } from 'inversify';
+import { Command } from '../../decorator';
+import { GitService } from '../../services/GitService';
+import { SparoProfileService } from '../../services/SparoProfileService';
+
+import type { Argv, ArgumentsCamelCase } from 'yargs';
+import type { ICommand } from './base';
+import type { TerminalService } from '../../services/TerminalService';
+
+export interface IPullCommandOptions {
+  branch?: string;
+  remote?: string;
+  profile?: string[];
+  addProfile?: string[];
+}
+
+@Command()
+export class PullCommand implements ICommand<IPullCommandOptions> {
+  public cmd: string = 'pull [remote] [branch]';
+  public description: string = 'pull changes from remote branch to local';
+
+  @inject(GitService) private _gitService!: GitService;
+  @inject(SparoProfileService) private _sparoProfileService!: SparoProfileService;
+
+  public builder(yargs: Argv<{}>): void {
+    /**
+     * sparo pull [remote] [branch] --profile <profile_name> --add-profile <profile_name> --no-profile
+     */
+    yargs
+      .positional('remote', { type: 'string' })
+      .positional('branch', { type: 'string' })
+      .string('remote')
+      .string('branch')
+      .boolean('full')
+      .array('profile')
+      .default('profile', [])
+      .array('add-profile')
+      .default('add-profile', []);
+  }
+
+  public handler = async (
+    args: ArgumentsCamelCase<IPullCommandOptions>,
+    terminalService: TerminalService
+  ): Promise<void> => {
+    const { _gitService: gitService, _sparoProfileService: sparoProfileService } = this;
+    const { terminal } = terminalService;
+
+    terminal.writeDebugLine(`got args in pull command: ${JSON.stringify(args)}`);
+    const pullArgs: string[] = ['pull'];
+
+    const { branch, remote } = args;
+
+    if (branch && remote) {
+      pullArgs.push(remote, branch);
+    }
+
+    const { isNoProfile, profiles, addProfiles } = await sparoProfileService.preprocessProfileArgs({
+      profilesFromArg: args.profile ?? [],
+      addProfilesFromArg: args.addProfile ?? []
+    });
+
+    // invoke native git pull command
+    gitService.executeGitCommand({ args: pullArgs });
+
+    // check whether profile exist in local branch
+    if (!isNoProfile) {
+      const targetProfileNames: Set<string> = new Set([...profiles, ...addProfiles]);
+      const nonExistProfileNames: string[] = [];
+      for (const targetProfileName of targetProfileNames) {
+        if (!this._sparoProfileService.hasProfileInFS(targetProfileName)) {
+          nonExistProfileNames.push(targetProfileName);
+        }
+      }
+
+      if (nonExistProfileNames.length) {
+        const { branch } = gitService.getRepoInfo();
+        throw new Error(
+          `Pull failed. The following profile(s) are missing in local branch "${branch}": ${Array.from(
```

Comment:
I think it's a little different compared with `checkout` command. 

For `pull` command, which is the combination of `git fetch` and `git merge/rebase`, it has the meaning of "update my local code". 

Here comes the core question, should we check profile existence based on old code? Or should we check profile existence based on updated code?

For example, if someone else add a new profile in my branch and I want to pull that changes and use that profile, I would probably run 
```shell
sparo pull --profile <new_profile_name>
```
If we check before "git pull" actually happens, we would never get that new profile.

------
> That's what we want to avoid since we don't have an answer to the next step to handle this incompleted state.

I think there are two options:
1. If sparse checkout failed and run into incomplete state, we could run `sparo checkout` or `sparo pull` to trigger sparse checkout again and fix that incomplete state.
2. Record current git commit before pull, if sparse checkout failed and run into incomplete state, use git reset to restore code changes (I would personally don't recommend to do git reset, as it's risky and might lose local code changes)
      * Show old commit and let user manually do git reset.
      * Automatically reset to old commit in sparo error catch logic.